# HR data analysis (part 1)

This is not a project with a story or a business task. I have specifically searched for an HR dataset to showcase my ability in mySQL and Tableau. I wanted to create a dynamic dashboard from scratch using an unknown __[dataset](https://data.world/markbradbourne/rwfd-real-world-fake-data/workspace/file?filename=Human+Resources.csv)__. Including getting familiar with the file, searching ways to explore and of cource, checking the logic and the consistency. Good thing on this topic is that everyone has a basic knowledge on gender distribution, race representation, work experience and other workplace characteristics - so we can concentrate on the main things.

### About the data

Data has been downloaded and stored locally and uploaded into Google Drive as a back-up.

As usual, we will treat this dataset as it's ROCCC:
- Reliable, Original and Cited: complete time series, which are accurate and non-bias. We will treat them as second-party datasets from a reliable organization.
- Current and Comprehensive: we consider them as regularly refreshed datasets, which are appropriate and will enable us to answer the business questions. 

We have the following columns:
- __id:__ unique employee id used as primary key
- __first_name:__ employee's first name
- __last_name:__ employee's last name
- __birthdate:__ the date of birth
- __gender:__ only categorical variable representing two genders: male or female (or Non-Conforming)
- __race:__ employee diversity
- __department:__ company's various divisions
- __jobtitle:__ employee's current jobtitle (including role level if applicable)
- __location:__ location type of work: remote or headquarter
- __hire_date:__ the date when the employee was hired
- __termdate:__ the date when the employee was let go 
- __location_city:__ the city of work
- __location_state:__ the state of work

### Setting up a business task

My goal is to create meaningful insights out of this company dataset.

1. What are the company values?
2. How they represent diversity and inclusion?
3. Is this an "equal opportunity" company?

### Exploratory data analysis

You can find the original code __[here](https://github.com/bettybuilds/HRdata/blob/main/hrdata_dataanalysis.sql)__ which was written in the mySQL Workbench. It contains the data manipulation and cleaning part. In this article I'm going to show only the EDA.
To be visually more aesthetic, I wanted it to be in a Jupyter notebook, but this time I'm going to use SQL Magic extension.

Please note that I'm going to limit the results on the website. If you would like to, you can download the csv and the script from the repository.

In [1]:
%load_ext sql

Loading in the dataset:

In [3]:
%sql mysql+mysqldb://root:<password>@localhost/dataset

Let's check how many employees we have by jobtitle:

In [4]:
%%sql

SELECT jobtitle, COUNT(jobtitle) AS nu_employee
FROM hrdata
GROUP BY jobtitle
ORDER BY nu_employee DESC
LIMIT 10;

 * mysql+mysqldb://root:***@localhost/dataset
10 rows affected.


[('Research Assistant II', 695),
 ('Business Analyst', 648),
 ('Human Resources Analyst II', 557),
 ('Research Assistant I', 492),
 ('Account Executive', 465),
 ('Data Visualization Specialist', 425),
 ('Staff Accountant I', 411),
 ('Human Resources Analyst', 380),
 ('Software Engineer I', 366),
 ('Systems Administrator I', 349)]

Now lets see what jobs we have by department:

In [5]:
%%sql

SELECT DISTINCT department, jobtitle
FROM hrdata
ORDER BY department
LIMIT 15;

 * mysql+mysqldb://root:***@localhost/dataset
15 rows affected.


[('Accounting', 'Accountant I'),
 ('Accounting', 'Accountant II'),
 ('Accounting', 'Accountant III'),
 ('Accounting', 'Accountant IV'),
 ('Accounting', 'Accounting Assistant I'),
 ('Accounting', 'Accounting Assistant II'),
 ('Accounting', 'Accounting Assistant III'),
 ('Accounting', 'Accounting Assistant IV'),
 ('Accounting', 'Actuary'),
 ('Accounting', 'Administrative Assistant I'),
 ('Accounting', 'Administrative Assistant II'),
 ('Accounting', 'Administrative Assistant III'),
 ('Accounting', 'Administrative Assistant IV'),
 ('Accounting', 'Administrative Officer'),
 ('Accounting', 'Budget/Accounting Analyst I')]

It seems like we have role levels for many jobs.
Would be interesting to check what is the average experience an employee need for each role:

In [6]:
%%sql

SELECT jobtitle, AVG(experience) AS avg_exp
FROM hrdata
GROUP BY jobtitle
ORDER BY jobtitle
LIMIT 15;

 * mysql+mysqldb://root:***@localhost/dataset
15 rows affected.


[('Account Coordinator', Decimal('6.0000')),
 ('Account Executive', Decimal('12.3613')),
 ('Account Manager', Decimal('11.2647')),
 ('Accountant I', Decimal('12.6800')),
 ('Accountant II', Decimal('12.0625')),
 ('Accountant III', Decimal('12.6867')),
 ('Accountant IV', Decimal('11.8750')),
 ('Accounting Assistant I', Decimal('13.0000')),
 ('Accounting Assistant II', Decimal('11.7531')),
 ('Accounting Assistant III', Decimal('12.2500')),
 ('Accounting Assistant IV', Decimal('13.5610')),
 ('Actuary', Decimal('12.5185')),
 ('Administrative Assistant', Decimal('11.7297')),
 ('Administrative Assistant I', Decimal('13.4754')),
 ('Administrative Assistant II', Decimal('11.8372'))]

We can see some inconsistency in the dataset as the experience in the company doesn't reflect the role level.
For example, if we look the first group with separeted role levels (the Accountants), we can see that there
is an average 11 or 12 years of experience for each of the four role levels.
This means that the company is prefering external hiring instead of promoting employees.

What about the gender distribution on these role levels?

In [7]:
%%sql

SELECT gender, jobtitle, AVG(experience) AS avg_exp
FROM hrdata
WHERE gender != 'Non-Conforming'
GROUP BY gender, jobtitle
ORDER BY avg_exp DESC
LIMIT 15;

 * mysql+mysqldb://root:***@localhost/dataset
15 rows affected.


[('Female', 'Engineer II', Decimal('21.0000')),
 ('Female', 'Engineer IV', Decimal('17.0000')),
 ('Male', 'Office Assistant I', Decimal('16.6667')),
 ('Female', 'Director of Sales', Decimal('16.6000')),
 ('Female', 'Sales Representative', Decimal('16.5000')),
 ('Female', 'Statistician II', Decimal('16.2500')),
 ('Male', 'Statistician II', Decimal('16.0000')),
 ('Male', 'Research Assistant III', Decimal('15.4000')),
 ('Male', 'Human Resources Assistant III', Decimal('15.3333')),
 ('Female', 'Research Assistant III', Decimal('15.2857')),
 ('Male', 'Analog Circuit Design manager', Decimal('15.2273')),
 ('Female', 'Human Resources Manager', Decimal('15.2000')),
 ('Female', 'Compensation Analyst', Decimal('15.0000')),
 ('Female', 'Statistician III', Decimal('15.0000')),
 ('Male', 'Assistant Professor', Decimal('15.0000'))]

There are slightly more female at the top of the list, which means that the highest years of experience
in the company are currently owned by females.
I'm not able to detect anything else for the role levels with this pivot.

In [9]:
%%sql

SELECT gender, AVG(experience) AS avg_exp, COUNT(gender) AS nu_gender
FROM hrdata
WHERE gender != 'Non-Conforming'
GROUP BY gender
ORDER BY avg_exp;

 * mysql+mysqldb://root:***@localhost/dataset
2 rows affected.


[('Male', Decimal('12.2417'), 10472), ('Female', Decimal('12.2695'), 9540)]

This is incorrect, because the order by function causes disturbance in calculation.
We should use the total count for calculating the average.

In [10]:
%%sql

SELECT COUNT(gender) AS total_count
FROM hrdata
WHERE gender != 'Non-Conforming';

 * mysql+mysqldb://root:***@localhost/dataset
1 rows affected.


[(20012,)]

The total count is 20,012.

In [11]:
%%sql

SELECT gender, SUM(experience)/'21609' AS gender_exp
FROM hrdata
WHERE gender != 'Non-Conforming'
GROUP BY gender;

 * mysql+mysqldb://root:***@localhost/dataset
2 rows affected.


[('Male', 5.932481836271924), ('Female', 5.416770789948632)]

So we can say that on average (based on the total employee count where the gender info is avaiable), the
males have more years of experience than the females.
We can also see that currently there are almost 1,000 more males working at the company. So the gender
distribution (from the avaiable data) doesn't seem to be bad.

What about the age?

In [12]:
%%sql

SELECT gender, age, COUNT(id) AS nu_employee
FROM hrdata
WHERE gender != 'Non-Conforming'
GROUP BY gender, age
ORDER BY age, nu_employee DESC
LIMIT 15;

 * mysql+mysqldb://root:***@localhost/dataset
15 rows affected.


[('Female', 20, 110),
 ('Male', 20, 97),
 ('Male', 21, 286),
 ('Female', 21, 262),
 ('Male', 22, 350),
 ('Female', 22, 264),
 ('Male', 23, 285),
 ('Female', 23, 252),
 ('Male', 24, 267),
 ('Female', 24, 234),
 ('Male', 25, 277),
 ('Female', 25, 270),
 ('Male', 26, 294),
 ('Female', 26, 265),
 ('Male', 27, 267)]

The count of employees grouped by age and gender could be a very interesting data for visualiation, which
we will do later. With bare eyes I don't detect any significant discrepancy between the genders.

In [13]:
%%sql

SELECT gender, age, AVG(experience) AS avg_exp
FROM hrdata
WHERE gender != 'Non-Conforming'
GROUP BY gender, age
ORDER BY age, avg_exp DESC
LIMIT 15;

 * mysql+mysqldb://root:***@localhost/dataset
15 rows affected.


[('Male', 20, Decimal('12.7732')),
 ('Female', 20, Decimal('11.6727')),
 ('Male', 21, Decimal('12.5315')),
 ('Female', 21, Decimal('11.9504')),
 ('Female', 22, Decimal('12.6061')),
 ('Male', 22, Decimal('12.5400')),
 ('Female', 23, Decimal('12.5000')),
 ('Male', 23, Decimal('12.3860')),
 ('Male', 24, Decimal('12.8464')),
 ('Female', 24, Decimal('12.0769')),
 ('Female', 25, Decimal('12.5741')),
 ('Male', 25, Decimal('12.5451')),
 ('Male', 26, Decimal('12.8265')),
 ('Female', 26, Decimal('12.6868')),
 ('Male', 27, Decimal('12.6067'))]

If we add the average experience we can see that the dataset is faulty in this aspect.
As we discussed at the very beginning (you can find the whole script __[here](https://github.com/bettybuilds/HRdata/blob/main/hrdata_dataanalysis.sql)__), unfortunately the creator of the dataset forgot
to add more years between the birthdate and the hire_date.

##### However, this fact won't stop us to make the most of our dataset, so let's continue the exporation!

We can check the distribution of the race as well, to see if there are any race inequalities.

In [14]:
%%sql

SELECT DISTINCT race FROM hrdata;


 * mysql+mysqldb://root:***@localhost/dataset
7 rows affected.


[('Hispanic or Latino',),
 ('White',),
 ('Black or African American',),
 ('Asian',),
 ('Two or More Races',),
 ('American Indian or Alaska Native',),
 ('Native Hawaiian or Other Pacific Islander',)]

In [15]:
%%sql

SELECT COUNT(race) FROM hrdata;

 * mysql+mysqldb://root:***@localhost/dataset
1 rows affected.


[(20574,)]

There are no missing data here, so it means that our total count is the row number: 20,574.

In [16]:
%%sql

SELECT race, (COUNT(race) / '22214') * 100 AS distr_race
FROM hrdata
GROUP BY race
ORDER BY distr_race DESC;

 * mysql+mysqldb://root:***@localhost/dataset
7 rows affected.


[('White', 26.41127217070316),
 ('Two or More Races', 15.161609795624381),
 ('Black or African American', 15.103088142612767),
 ('Asian', 14.850994868101196),
 ('Hispanic or Latino', 10.4258575672999),
 ('American Indian or Alaska Native', 5.559557036103358),
 ('Native Hawaiian or Other Pacific Islander', 5.104888808859278)]

Almost ~29% of the company are Whites, another bigger bites are Multiracials, Black/African Americans
and Asians with ~16% each.
Hispanic/Latinos are significantly less represented in the company with a 11%. Meanwhile the
American Indian/Alaska natives, the native Hawaiians or any other Pacific Isnlanders are present only ~6%.

We can check the data regarding the experience:

In [17]:
%%sql

SELECT race, (COUNT(race) / '22214') * 100 AS distr_race, AVG(experience) AS avg_exp
FROM hrdata
GROUP BY race
ORDER BY avg_exp DESC;

 * mysql+mysqldb://root:***@localhost/dataset
7 rows affected.


[('Native Hawaiian or Other Pacific Islander', 5.104888808859278, Decimal('12.4048')),
 ('Black or African American', 15.103088142612767, Decimal('12.3133')),
 ('Two or More Races', 15.161609795624381, Decimal('12.2824')),
 ('Hispanic or Latino', 10.4258575672999, Decimal('12.2785')),
 ('Asian', 14.850994868101196, Decimal('12.2543')),
 ('White', 26.41127217070316, Decimal('12.1807')),
 ('American Indian or Alaska Native', 5.559557036103358, Decimal('12.1498'))]

There is no significant discrepancy in the experience between the different races.
Black/African Americans usually have slightly more experience than most of the company, but the
Native Hawaiians/Other Pacific Isnalders are at the top.

We can check the jobtitles for each race in a pivot as well, but it's too much information. We should check
this in a chart later.

In [18]:
%%sql

SELECT race, jobtitle, AVG(experience) AS avg_exp
FROM hrdata
GROUP BY race, jobtitle
ORDER BY avg_exp DESC
LIMIT 15;

 * mysql+mysqldb://root:***@localhost/dataset
15 rows affected.


[('American Indian or Alaska Native', 'Developer II', Decimal('22.0000')),
 ('Hispanic or Latino', 'Web Designer IV', Decimal('22.0000')),
 ('Native Hawaiian or Other Pacific Islander', 'Office Assistant I', Decimal('22.0000')),
 ('American Indian or Alaska Native', 'Automation Specialist IV', Decimal('22.0000')),
 ('Native Hawaiian or Other Pacific Islander', 'Software Engineer IV', Decimal('21.0000')),
 ('White', 'Sales Associate', Decimal('21.0000')),
 ('Asian', 'Office Assistant II', Decimal('21.0000')),
 ('Native Hawaiian or Other Pacific Islander', 'Senior Sales Associate', Decimal('20.5000')),
 ('Native Hawaiian or Other Pacific Islander', 'VP Accounting', Decimal('20.5000')),
 ('Native Hawaiian or Other Pacific Islander', 'Attorney', Decimal('20.0000')),
 ('American Indian or Alaska Native', 'Support Staff II', Decimal('20.0000')),
 ('Native Hawaiian or Other Pacific Islander', 'Mechanical Systems Engineer', Decimal('20.0000')),
 ('Black or African American', 'Research Assistant IV', Decimal('20.0000')),
 ('American Indian or Alaska Native', 'Internal Auditor', Decimal('19.6667')),
 ('Native Hawaiian or Other Pacific Islander', 'Research Assistant III', Decimal('19.5000'))]

We can also check the race distribution on each location by city:

In [19]:
%%sql

SELECT location_city, race, COUNT(race) AS nu_race
FROM hrdata
GROUP BY location_city, race
ORDER BY location_city, nu_race DESC
LIMIT 15;

 * mysql+mysqldb://root:***@localhost/dataset
15 rows affected.


[('Akron', 'White', 43),
 ('Akron', 'Two or More Races', 20),
 ('Akron', 'Black or African American', 17),
 ('Akron', 'Hispanic or Latino', 15),
 ('Akron', 'Asian', 14),
 ('Akron', 'American Indian or Alaska Native', 8),
 ('Akron', 'Native Hawaiian or Other Pacific Islander', 5),
 ('Allentown', 'Asian', 5),
 ('Allentown', 'White', 4),
 ('Allentown', 'Two or More Races', 3),
 ('Allentown', 'Black or African American', 3),
 ('Allentown', 'Hispanic or Latino', 3),
 ('Allentown', 'Native Hawaiian or Other Pacific Islander', 1),
 ('Anderson', 'White', 6),
 ('Anderson', 'Black or African American', 5)]

.. and by state.

In [20]:
%%sql

SELECT location_state, race, COUNT(race) AS nu_race
FROM hrdata
GROUP BY location_state, race
ORDER BY location_state, nu_race DESC
LIMIT 15;

 * mysql+mysqldb://root:***@localhost/dataset
15 rows affected.


[('Illinois', 'White', 226),
 ('Illinois', 'Two or More Races', 156),
 ('Illinois', 'Asian', 131),
 ('Illinois', 'Black or African American', 117),
 ('Illinois', 'Hispanic or Latino', 91),
 ('Illinois', 'American Indian or Alaska Native', 47),
 ('Illinois', 'Native Hawaiian or Other Pacific Islander', 42),
 ('Indiana', 'White', 186),
 ('Indiana', 'Asian', 109),
 ('Indiana', 'Black or African American', 109),
 ('Indiana', 'Two or More Races', 106),
 ('Indiana', 'Hispanic or Latino', 55),
 ('Indiana', 'Native Hawaiian or Other Pacific Islander', 52),
 ('Indiana', 'American Indian or Alaska Native', 28),
 ('Kentucky', 'White', 121)]

There is one more interesting data we haven't check yet: remote vs office.

In [21]:
%%sql

SELECT DISTINCT location FROM hrdata;

 * mysql+mysqldb://root:***@localhost/dataset
2 rows affected.


[('Headquarters',), ('Remote',)]

First we can check what is the distribution of the employees:

In [22]:
%%sql

SELECT COUNT(id) AS nu_employee, location
FROM hrdata
GROUP BY location;

 * mysql+mysqldb://root:***@localhost/dataset
2 rows affected.


[(15469, 'Headquarters'), (5105, 'Remote')]

Currently the preferred working place is in the office.

What is the distribution between the genders?

In [23]:
%%sql

SELECT COUNT(id) AS nu_employee, location, gender
FROM hrdata
WHERE gender != 'Non-Conforming'
GROUP BY location, gender
ORDER BY nu_employee DESC;

 * mysql+mysqldb://root:***@localhost/dataset
4 rows affected.


[(7863, 'Headquarters', 'Male'),
 (7182, 'Headquarters', 'Female'),
 (2609, 'Remote', 'Male'),
 (2358, 'Remote', 'Female')]

Nevertheless, we can check the location for race:

In [24]:
%%sql

SELECT COUNT(id) / 22214 * 100 AS nu_employee, location, race
FROM hrdata
GROUP BY location, race
ORDER BY location, nu_employee DESC
LIMIT 15;

 * mysql+mysqldb://root:***@localhost/dataset
14 rows affected.


[(Decimal('19.7038'), 'Headquarters', 'White'),
 (Decimal('11.4477'), 'Headquarters', 'Black or African American'),
 (Decimal('11.3577'), 'Headquarters', 'Two or More Races'),
 (Decimal('11.2587'), 'Headquarters', 'Asian'),
 (Decimal('7.9454'), 'Headquarters', 'Hispanic or Latino'),
 (Decimal('4.2091'), 'Headquarters', 'American Indian or Alaska Native'),
 (Decimal('3.7139'), 'Headquarters', 'Native Hawaiian or Other Pacific Islander'),
 (Decimal('6.7075'), 'Remote', 'White'),
 (Decimal('3.8039'), 'Remote', 'Two or More Races'),
 (Decimal('3.6554'), 'Remote', 'Black or African American'),
 (Decimal('3.5923'), 'Remote', 'Asian'),
 (Decimal('2.4804'), 'Remote', 'Hispanic or Latino'),
 (Decimal('1.3910'), 'Remote', 'Native Hawaiian or Other Pacific Islander'),
 (Decimal('1.3505'), 'Remote', 'American Indian or Alaska Native')]

It's good to see here that both in the gender and in the race distribution percentages we can recognize the location distributions, as that means the company doesn't decide on the working place based on gender or race.

From my experience, the jobtitle (and maybe the role level) would be the key factor, for which wouldn't be the
best idea to visualize in a pivot, but in a chart.

In [25]:
%%sql

SELECT location, jobtitle
FROM hrdata
GROUP BY location, jobtitle
ORDER BY location
LIMIT 15;

 * mysql+mysqldb://root:***@localhost/dataset
15 rows affected.


[('Headquarters', 'Account Coordinator'),
 ('Headquarters', 'Account Executive'),
 ('Headquarters', 'Account Manager'),
 ('Headquarters', 'Accountant I'),
 ('Headquarters', 'Accountant II'),
 ('Headquarters', 'Accountant III'),
 ('Headquarters', 'Accountant IV'),
 ('Headquarters', 'Accounting Assistant I'),
 ('Headquarters', 'Accounting Assistant II'),
 ('Headquarters', 'Accounting Assistant III'),
 ('Headquarters', 'Accounting Assistant IV'),
 ('Headquarters', 'Actuary'),
 ('Headquarters', 'Administrative Assistant'),
 ('Headquarters', 'Administrative Assistant I'),
 ('Headquarters', 'Administrative Assistant II')]

With naked eyes it's really hard to detect patterns and apart from this fact, there is no guarantee that the
jobtitle would reflect the location since the best case scenario would be to let the employee decide where to work.
We will check later on this in a chart.

#### Feedback, bug reports, and comments are not only welcome, but strongly encouraged!